### ISSUE TITLE
Weird { _: [], '$0': 'test/test.js', test: 'TAP', message: undefined } outside of test

### ISSUE THREAD

**Issue by philkunz** [tapjs/tapjs#365]
Hi there I have like a really weird output

```
{ _: [], '$0': 'test/test.js', test: 'TAP', message: undefined }
```
The error is not from any test, yet generated by tap. But even catching errors with catch(tap.threw) doesn't give any insight:

I don't get this error when I remove this test

```javascript
tap.test('should start parsing the CLI input', async () => {
  smartCliTestObject.startParse()
  return await expect(smartCliTestObject.parseStarted.promise).to.eventually.be.fulfilled
}).catch(tap.threw)
```

However the test itself turns out green. There are no subsequent tests that depend on it in any way, nor does `smartCliTestObject.startParse()`itself throw anything.

build log is here:
https://gitlab.com/pushrocks/smartcli/builds/14833419

Any ideas?

**Comment by philkunz** [comment]
So tap parser says one failed test, that I can't attribute to any test defined by me when running that last test that succeeds... Something feels off somehow. Didn't have any problems with mocha here...

**Comment by isaacs** [comment]
So, for starters, this is kinda weird to me, probably not what you're intending:

```js
tap.test('should not have executed yet', async () => {
  return expect(hasExecuted).to.be.false
}).catch(tap.threw)
```

You're putting a `.catch()` on the promise returned by `tap.test()`, but that'll definitely never ever reject, so it's just extra code that does nothing.  Best to leave it off.  (Throws are already handled by tap as test failures -- effectively, you're putting a .catch() on a promise that already that exact one.)

The output seems to indicate that you're throwing (or rejecting a promise with) an object which is very much not an Error.  Looks to me like a parsed command-line?  Maybe it's the `expect` thing misinterpreting the result as an error?

```
  1) 
{ _: [], '$0': 'test/test.js', test: 'TAP', message: undefined }
```

The spec output formatter isn't my favorite.  Not much to help us there, really.  I recommend either the `-Rclassic` default, or `-Rtap` to dump the raw output.  You can also do `-o some-file.tap` and then cat that after the test run if it fails to see all the raw gory details.

Let's look at the test that's causing it:

```js
tap.test('should start parsing the CLI input', async () => {
  smartCliTestObject.startParse()
  return await expect(smartCliTestObject.parseStarted.promise).to.eventually.be.fulfilled
})
```

You don't really have to do this async and return-await stuff.  If you just wanna verify that it resolves, then this works just fine:

```js
tap.test('should start parsing the CLI input', async () => {
  smartCliTestObject.startParse()
  await smartCliTestObject.parseStarted.promise
})
```

Tap will detect the Promise returned by the async function, and pass the test when it resolves, or fail the test when it rejects.  (Async functions return a promise that resolves at the end of the run-to-completion after their last awaited promise resolves.)

It's hard to have any further insight, given the number of other things in play.  If you can generate the weird output using tap directly, I'd love a minimal repro case.

**Comment by philkunz** [comment]
First of all: Thank you for responding on a Sunday. :) I really do appreciate this.

Good tip with the .catch(tap.threw). Does make sense to me.

The reason for...

```javascript
return await expect(smartCliTestObject.parseStarted.promise).to.eventually.be.fulfilled
```

... is that I like the consistent handling of expect for deciding between right and wrong.

`_: []` Looks like CLI indeed. Its just weird why there is any tap in there... :D I'll keep investigating and see if I can make a minimal repro case.

Have a nice Sunday

Regards
Phil

**Comment by philkunz** [comment]
found my issue. I was rejecting a promise for all commands that did not trigger :) I originally implemented this behaviour to do stuff in case a command is not specified.

I did so with "q" back in the time when promises that were rejected but not catched did so silently without causing any trouble... Now I switched to native promises (and from mocha to tap) and this behaviour popped up.

So not tap's fault. Sorry for your taking some of your time. mea culpa :)
